imp: exec: send SIGKILL to cgroup/children on SIGUSR1

grondo opened the pull request:
This PR updates the IMP to translate SIGUSR1 to SIGKILL for `flux-imp exec` as defined in RFC 15.

Additionally, if possible, the IMP will signal all members of its cgroup (besides itself) when delivering SIGKILL.

Finally, to better handle returning the proper wait status to _its_ parent, when the child of the IMP is terminated by a signal, the IMP raises that same signal to itself.

This is WIP pending real system testing.

grondo commented:
Hm, readthedocs build seems to have been broken at some point as well. Will have to look into that.

garlick commented on src/imp/cgroup.c:
```diff
@@ -0,0 +1,172 @@
+                          "%s%s",
+                          cgroup->mount_dir,
+                          relpath);
+            if ((n > 0) && (n < len))
+                rc = 0;
```
I don't think n <= 0 is a possible failure.  Then if only checking for one condition, the n (re-)assignment could b e avoided.

garlick commented on src/imp/cgroup.c:
```diff
@@ -0,0 +1,172 @@
+        char *subsys = strchr (line, ':');
+        if ((nl = strchr (line, '\n')))
+            *nl = '\0';
+        if (subsys == NULL || *(++subsys) == '\0'
+            || !(relpath = strchr (subsys, ':')))
```
style: split to one conditional per line?

grondo replied:
Sorry, this code is actually copied from existing code. I'll clean that up.

garlick commented on src/imp/cgroup.c:
```diff
@@ -34,6 +34,7 @@
 #ifndef CGROUP_SUPER_MAGIC
```
commit message: s/delivery/deliver/

garlick commented on src/imp/cgroup.c:
```diff
@@ -168,5 +169,41 @@ struct cgroup_info *cgroup_info_create (void)
+    int count = 0;
+    int rc = 0;
+    int saved_errno = 0;
+    char path [PATH_MAX+15]; /* cgroup->path + "/cgroup.procs" */
```
Did you mean `PATH_MAX+1`?

grondo replied:
hm. cgroup->path is a buffer of size `PATH_MAX`. We're going to write 14 more characters so we need at least that much space or the compiler complains. Happy to do this another way.

garlick replied:
Oh right, I hate that.  Oh well.

garlick commented on src/imp/cgroup.c:
```diff
@@ -168,5 +169,41 @@ struct cgroup_info *cgroup_info_create (void)
+    unsigned long child;
+    pid_t current_pid = getpid ();
+
+    (void) snprintf (path, sizeof (path), "%s/cgroup.procs", cgroup->path);
```
Check for overflow?

grondo replied:
Here we know we've added enough space to the buffer so it can't overflow. I guess `sprintf` could be used with a note in that case?

garlick commented on src/imp/signals.c:
```diff
@@ -107,5 +108,16 @@ void imp_setup_signal_forwarding (struct imp_state *imp)
 
+void imp_raise (int sig)
+{
+    signal (sig, SIG_DFL);
+    if (raise (sig) == 0)
+        pause ();
```
Oh, nice!  I hadn't even thought about the fact that the wait status would not have WIFSIGNALED true etc. in this case.

garlick commented on src/test/docker/docker-run-checks.sh:
```diff
@@ -177,7 +177,7 @@ else
     docker run --rm \
```
commit message: /expclitly/explicitly/

garlick reviewed (approved):
Great!

I have some comments but all superficial and/or optional.

grondo commented:
As part of this work I switched the cgroup controller that the IMP checks for from `name=systemd` to `pids`, since the latter can actually be used in testing. I had thought this was working, but after further investigation, that isn't quite right and the `pids` controller is unchanged for processes in a job managed by sdexec.

I think the correct thing (after perusing [cgroups(7)](https://man7.org/linux/man-pages/man7/cgroups.7.html)) is to use the `0::` line in `/proc/self/cgroup`, which I _think_ corresponds to the cgroups v2 hierarchy.

I'll have to modify this PR to make that change.

grondo commented:
Ok, I think I've fixed the cgroup discovery here. The code is somewhat simplified, at startup the IMP:

 - checks if either `/sys/fs/cgroup` or `/sys/fs/cgroup/unified` are mounted as cgroup2
 - if not then check for a legacy (cgroup v1) mount at `/sys/fs/cgroup/systemd`
 - grab the current relative cgroup path for either case from `/proc/self/cgroup` (logic to find the correct one is now fixed)
 
I also had to add a workaround to remove leading `/..` from the relative cgroup path, which seems to pop up when run within a container (the relative path is relative to the container mount point on the host I think)

garlick commented:
I grabbed your snapshot from last night and put it on my test cluster.  This morning I made some minimal changes to job-exec on my `issue#6406` branch to send SIGUSR1 instead of imp-kill when shell termination methods fail.  Then I ran a test on my system by starting a long running job, sending the flux-shell a SIGSTOP, then canceling the job.  Looks good!  Flux on rank 0 logged:
```
[  +5.002092] job-exec[0]: Sending SIGKILL to job ƒ3L9MFPYEKZ
[ +10.005491] job-exec[0]: Sending SIGKILL to job ƒ3L9MFPYEKZ
[ +15.005766] job-exec[0]: Sending SIGKILL to job ƒ3L9MFPYEKZ
[ +20.005115] job-exec[0]: Sending SIGKILL to job ƒ3L9MFPYEKZ
[ +25.005382] job-exec[0]: Sending SIGKILL to job ƒ3L9MFPYEKZ
[ +25.005415] job-exec[0]: Sending SIGUSR1 to IMP for job ƒ3L9MFPYEKZ
[ +25.169307] job-manager[0]: housekeeping: ƒ3L9MFPYEKZ started
[ +25.201629] job-list[0]: purged 1 inactive jobs
[ +25.203373] job-manager[0]: purged 1 inactive jobs
[Nov01 08:43] job-manager[0]: housekeeping: ƒ3L9MFPYEKZ complete
```
On the compute node, systemd logged
```
Nov 01 08:41:05 picl1 systemd[796]: Starting imp-shell-1-f3L9MFPYEKZ.service - User workload...
Nov 01 08:41:05 picl1 systemd[796]: Started imp-shell-1-f3L9MFPYEKZ.service - User workload.
Nov 01 08:42:32 picl1 systemd[796]: imp-shell-1-f3L9MFPYEKZ.service: Sent signal SIGTERM to main process 907 (flux-imp) on client request.
Nov 01 08:42:57 picl1 systemd[796]: imp-shell-1-f3L9MFPYEKZ.service: Sent signal SIGUSR1 to main process 907 (flux-imp) on client request.
Nov 01 08:42:57 picl1 systemd[796]: imp-shell-1-f3L9MFPYEKZ.service: Main process exited, code=killed, status=9/KILL
Nov 01 08:42:57 picl1 systemd[796]: imp-shell-1-f3L9MFPYEKZ.service: Failed with result 'signal'.
```

grondo commented:
This morning I added one fix here: `cgroup_kill()` was always sending `SIGKILL` instead of the `sig` parameter. Fixed.

garlick reviewed (approved):
LGTM!

grondo commented:
Great, thanks! I'll set MWP. I'll post a PR to have the IMP wait until its cgroup is empty in a bit.
